addons/mass_mailing/models/mailing.py

```diff
@@ -260,15 +260,30 @@ def _compute_statistics(self):
             self.browse(row.pop('mailing_id')).update(row)
 
     def _compute_next_departure(self):
-        cron_next_call = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo().nextcall
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
         str2dt = fields.Datetime.from_string
-        cron_time = str2dt(cron_next_call)
+        cron_time = str2dt(cron_job.nextcall)
+        cron_id = cron_job.id
+
+        # The cron worker usually clean up old cron.triggers after finishing their
+        # jobs, i.e older than now() thus fetched recordset should not be to large
+        now_datetime_buffer = fields.Datetime.now() + relativedelta(minutes=1)
+        now_datetime_str = fields.Datetime.to_string(now_datetime_buffer)
+        cron_triggers = self.env['ir.cron.trigger'].sudo().search(['&', ('cron_id', '=', cron_id), ('call_at', '<', now_datetime_str)])
+        most_recent_trigger_datetime = max([str2dt(tr.call_at) for tr in cron_triggers]) if cron_triggers else False
```

Comment:
you can use `order='call_at desc'` and `limit=1` as parameters for `search` so that all that is done by Postgres

Reply:
Also I think you meant this, right?
```suggestion
        cron_triggers = self.env['ir.cron.trigger'].sudo().search(['&', ('cron_id', '=', cron_id), ('call_at', '>', now_datetime_str)])
        most_recent_trigger_datetime = max([str2dt(tr.call_at) for tr in cron_triggers]) if cron_triggers else False
```

Reply:
From my debugging session, the compute gets called before the trigger is created, then after it's created. So the trigger call_at date should be slightly younger than now (added a buffer of 1 minute just to be sure). So the less than logic holds up, since what we want is to be sure that there is at least one trigger younger that now for mailing object where `schedule_date = False`.

Reply:
I'm not sure I get the point, here's my understanding:

- I give a schedule_date to a mailing
- Whenever the cron is triggered, if the schedule_date is past, the mailing is sent
- If my schedule date is before the next trigger, then what is relevant is actually the trigger.
- So for any mailing the next departure should be `max(min(next_trigger, next_cron), schedule_date)`
 with `next_trigger` being the next trigger after `max(now, schedule_date)`
 -> note that this is idealized as it could get a bit expensive to check the next trigger after each schedule_date if we need to look at all the triggers, and it probably won't bring much value.

That specific compute will run every time we fetch this record so it shouldn't be a big deal. But if there really is a microsecond discrepency issue, we can just add some negative padding to 'now' to be sure right? Though I'm not a fan.

Let me know if I got something wrong or if it's not clear :)

Reply:
This is indeed a way we could simplify it.
If we fetch the most recent trigger (younger than `now`), to treat all possible cases:
- user schedules a date in the past `we are 10 FEB, user puts 01FEB` (should this even be possible? because right now it is) : trigger is created for 01FEB. Since `schedule_date` gets set, we go to `max(mass_mailing.schedule_date, fields.datetime.now())`, as the trigger is way in the past, it should wake the the worker execute it now [ OUTPUT : fields.datetime.now()]
- user presses "send":  `schedule_date` stays `False` and trigger for `datetime.now()` is created. Here we could just go for `mass_mailing.next_departure = most_recent_trigger_datetime` [ OUTPUT : most_recent_trigger_datetime]
- user schedules a date in the future: `schedule_date`  is set and trigger gets created in the future (after `now`). Here `mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())` kicks in again and should give the close enough next_departure  [ OUTPUT : mass_mailing.schedule_date ]

(Not disagreeing with you but just giving my summary of the use cases. If I missed something point it out :)